hotspot/main.go

```diff
@@ -0,0 +1,264 @@
+// with a single table.
+func setupDatabase(dbURL string) (*sql.DB, error) {
+	parsedURL, err := url.Parse(dbURL)
+	if err != nil {
+		return nil, err
+	}
+	parsedURL.Path = "hot"
+
+	// Open connection to server and create a database.
+	db, err := sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		return nil, err
+	}
+	if _, err := db.Exec(createDatabaseStatement); err != nil {
+		return nil, err
+	}
+
+	// Allow a maximum of concurrency+1 connections to the database.
+	db.SetMaxOpenConns(*concurrency + 1)
+	db.SetMaxIdleConns(*concurrency + 1)
+
+	// Create the initial table for storing blocks.
+	if _, err := db.Exec(createTableStatement); err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+var usage = func() {
+	fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
+	fmt.Fprintf(os.Stderr, "  %s <db URL>\n\n", os.Args[0])
+	flag.PrintDefaults()
+}
+
+func main() {
+	flag.Usage = usage
+	flag.Parse()
+
+	dbURL := "postgresql://root@localhost:26257/photos?sslmode=disable"
+	if flag.NArg() == 1 {
+		dbURL = flag.Arg(0)
+	}
+
+	if *concurrency < 1 {
+		log.Fatalf("Value of 'concurrency' flag (%d) must be greater than or equal to 1", *concurrency)
+	}
+
+	if *writePercent < 0 || *writePercent > 100 {
+		log.Fatalf("Value of 'writePercent' flag (%d) must be between 0 and 100", *writePercent)
+	}
+
+	var db *sql.DB
+	{
+		var err error
+		for err == nil || *tolerateErrors {
+			db, err = setupDatabase(dbURL)
+			if err == nil {
+				break
+			}
+			if !*tolerateErrors {
+				log.Fatal(err)
+			}
+		}
+	}
+
+	lastNow := time.Now()
+	start := lastNow
+	var lastReads, lastWrites, lastTotal uint64
+	writers := make([]hotSpotWriter, *concurrency)
+
+	errCh := make(chan error)
+	var wg sync.WaitGroup
+	for i := range writers {
+		wg.Add(1)
+		writers[i] = newHotSpotWriter(db)
+		go writers[i].run(errCh, &wg)
+	}
+
+	var numErr int
+	tick := time.Tick(*outputInterval)
+	done := make(chan os.Signal, 3)
+	signal.Notify(done, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		wg.Wait()
+		done <- syscall.Signal(0)
+	}()
+
+	if *duration > 0 {
+		go func() {
+			time.Sleep(*duration)
+			done <- syscall.Signal(0)
+		}()
+	}
+
+	defer func() {
+		// Output results that mimic Go's built-in benchmark format.
+		elapsed := time.Since(start)
+		reads := atomic.LoadUint64(&readCount)
+		writes := atomic.LoadUint64(&writeCount)
+		total := reads + writes
+		fmt.Printf("%s\t%8d\t%12.1f ns/op\n",
+			*benchmarkName, total, float64(elapsed.Nanoseconds())/float64(total))
+	}()
+
+	for {
+		select {
+		case err := <-errCh:
+			numErr++
+			if !*tolerateErrors {
+				log.Fatal(err)
+			} else {
+				log.Print(err)
+			}
+			continue
+
+		case <-tick:
+			now := time.Now()
+			elapsed := time.Since(lastNow)
+			reads := atomic.LoadUint64(&readCount)
+			writes := atomic.LoadUint64(&writeCount)
+			total := reads + writes
+			fmt.Printf("%6s: reads:%6.1f/sec,\twrites:%6.1f/sec,\ttotal:%6.1f/sec",
```

Comment:
I think the tabular output of `allocsim` would be more readable here, though it doesn't matter too much.